### ISSUE TITLE
server: exception during startup isn't handled properly

### ISSUE THREAD

**Issue by hallazzang** [aio-libs/aiohttp#3899]
## Long story short

An exception thrown during startup isn't handled properly.
In fact, It can be seen only after shutting down the app.

## Expected behaviour

An exception during startup process should be raised immediately.

## Actual behaviour

Only after shutting down the app(by `KeyboardInterrupt`, etc.), the exception is raised.

```
$ python3.7 main.py
======== Running on http://0.0.0.0:8000 ========
(Press CTRL+C to quit)
^Cunhandled exception during asyncio.run() shutdown
task: <Task finished coro=<_run_app() done, defined at /home/ubuntu/.local/lib/python3.7/site-packages/aiohttp/web.py:261> exception=RuntimeError()>
Traceback (most recent call last):
  File "/home/ubuntu/.local/lib/python3.7/site-packages/aiohttp/web.py", line 348, in _run_app
    await asyncio.sleep(3600)  # sleep forever by 1 hour intervals
  File "/usr/lib/python3.7/asyncio/tasks.py", line 595, in sleep
    return await future
concurrent.futures._base.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ubuntu/.local/lib/python3.7/site-packages/aiohttp/web.py", line 350, in _run_app
    await runner.cleanup()
  File "/home/ubuntu/.local/lib/python3.7/site-packages/aiohttp/web_runner.py", line 222, in cleanup
    await self._cleanup_server()
  File "/home/ubuntu/.local/lib/python3.7/site-packages/aiohttp/web_runner.py", line 308, in _cleanup_server
    await self._app.cleanup()
  File "/home/ubuntu/.local/lib/python3.7/site-packages/aiohttp/web_app.py", line 403, in cleanup
    await self.on_cleanup.send(self)
  File "/home/ubuntu/.local/lib/python3.7/site-packages/aiohttp/signals.py", line 34, in send
    await receiver(*args, **kwargs)  # type: ignore
  File "main.py", line 16, in on_cleanup
    await app['background_task']
  File "main.py", line 7, in background_task
    raise RuntimeError
RuntimeError
```

(Here you can see that keyboard interrupt indicator(`^C`) on 4th line of the output)

## Steps to reproduce

Here's the code(`main.py`) I used to demonstrate this weird behavior.

```python
import asyncio

from aiohttp import web


async def background_task(app):
    raise RuntimeError


async def on_startup(app):
    app['background_task'] = asyncio.create_task(background_task(app))


async def on_cleanup(app):
    app['background_task'].cancel()
    await app['background_task']


def main():
    app = web.Application()
    app.on_startup.append(on_startup)
    app.on_cleanup.append(on_cleanup)

    web.run_app(app, port=8000)


if __name__ == '__main__':
    main()
```

## Your environment

- os: ubuntu 16.04.6 LTS on AWS EC2
- python version: 3.7.4
- aiohttp(server) version: 3.5.4

**Comment by atemate** [comment]
I think, the problem is with the test not with aiohttp. Try to change your `background_task` to print anything prior to raising an exception, and you'll see this message during app startup:
```
async def background_task(app):
    print("I'm here")
    raise RuntimeError
...
```
```
$ python create_task_app_bad.py 
I'm here
======== Running on http://0.0.0.0:8000 ========
(Press CTRL+C to quit)
^Cunhandled exception during asyncio.run() shutdown
task: <Task finished coro=<_run_app() done, defined at /usr/lib/python3.7/site-packages/aiohttp/web.py:261> exception=RuntimeError() created at /usr/lib/python3.7/asyncio/base_events.py:563>
...

During handling of the above exception, another exception occurred:

...
  File "create_task_app_bad.py", line 8, in background_task
    raise RuntimeError
RuntimeError
```

So the task `background_task` is scheduled, but the exception is not thrown out to the main execution until it is awaited directly or cancelled. In the proposed test code, the task is not awaited anywhere (note: only the coroutine `on_startup` is awaited during `web.run_app`, not `background_task`), but it is cancelled explicitly during the call of `on_cleanup`:
```
app['background_task'].cancel()
```

So I see it reasonable that the exception is thrown during the server shutdown.

Please note that since `on_startup` and `on_cleanup` are not connected anyhow, it's better to use `Application.cleanup_ctx` instead (see https://docs.aiohttp.org/en/stable/web_advanced.html#cleanup-context).

Thought I don't see this issue as a bug, I think since the questions on behaviour of `on_startup` and `on_cleanup` arise, we might want to update documentation by adding information on exception handling on these methods.

**Comment by asvetlov** [comment]
@ayushkovskiy is right.
Startup procedure is:
```
async def on_startup(app):
    app['background_task'] = asyncio.create_task(background_task(app))
```
It never fails: creation of a task for `background_task(app)` coroutine running is always succeded.
The task execution fails but it is another story that happens after startup procedure is finished.

To get task result you should call `await app['background_task']` which is actuaaly called in shutdown procedure.
That's why you see the exception there.

You code works as *expected*. If it doesn't satisfy your requirement -- write another code :)

**Comment by hallazzang** [comment]
Ok, now I understand what's going behind the scene. I've been admittedly misunderstanding `async`-related stuffs.

But I think the [documentation] should be updated, since anyone can experience this situation, not seeing any exception during startup process until shutting down the app, by following the documentation.

[documentation]: https://docs.aiohttp.org/en/stable/web_advanced.html#background-tasks

**Comment by asvetlov** [comment]
@ayushkovskiy suggested the doc update but I doubt if this is the right direction.
The initial problem is a misunderstanding of asyncio basics, not aiohttp specific problems.
Like we don't add a special warning for division by zero, the aiohttp docs don't teach how to write asyncio code in general but HTTP-related things only.

**Comment by hallazzang** [comment]
Ok then it's good. Thank you for your fast feedback
